### ISSUE TITLE
Tags for inserting basic input fields

### ISSUE THREAD

**Issue by ddoyen** [PremierLangage/Yggdrasil#38]
With @esandier, we suggest to use tags to insert basic input fields (radio, checkbox, drag&drop, matrix, math expression, jsxgraph, etc.) in the exercise `form`.

An input field would inserted in the `form` using a tag `{{input_name}}`. Then, the parameters of the input field would be defined in the body of the PL file in the namespace `input.name`.

Here is an example with a radio input.

~~~~
text ==
Quelle équipe la France a-t-elle battue en demi-finale de la Coupe du Monde 2018 ?
==

form ==
{{input_1}}
==

input.1.type = radio

input.1.shuffle = yes

input.1.choices ==
["Belgique",
"Brésil",
"Argentine",
"Uruguay"]
==

input.1.numsol = 0
~~~~

To create an input type, one would need to define:
- an html template
- (optional) a python script that processes the parameters
- (optional) a python script that processes the answer returned by the input field

Of course, the builder and the grader have to be modified.

Here are the advantages over a simple `form` template:
- multiple input fields in the form
- insertion of the input field in a more complex html code
- the python scripts allow an easy pre- and post-processing (e.g., shuffling the choices in a MCQ)

**Comment by ddoyen** [comment]
Some examples:

- `radio`
https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/radio.pl

- `checkbox`
https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/checkbox.pl

- `sortlist`
https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/sortlist.pl

- `mathexpr`
https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/mathexpr.pl

**Comment by nimdanor** [comment]
`sortlist.pl`
```python
before ==
import random
biglist=list(range(1,101))
choices=random.sample(biglist,4)
choices.sort()
==

text ==
Classer les nombres suivants par ordre croissant.
==

form ==
{{input_1}}
==

input.1.type = sortlist
input.1.choices = {{choices}}

evaluator ==
# FIXME c'est compliqué comme interface 
# answer['1']['eval']==True
# que pense tu de 
# score = input.1.evaluation()
# score = evaluation("1")
# un truc mieux 
if answer['1']['eval']==True:
    score=100
else:
    score=0
==
```

**Comment by ddoyen** [comment]
J'avais essayé de garder la syntaxe générique où toutes les réponses sont dans le dictionnaire answer.

Et comme je voulais disposer d'une réponse détaillée (un booléen qui dit si la liste est bien triée ou pas mais aussi la liste rentrée par l'élève), j'en étais arrivé naturellement à un dictionnaire dans le dictionnaire answer...

Mais effectivement, cette syntaxe est à discuter.

**Comment by esandier** [comment]
Après le wimsathon et discussion avec Dominique, je me demande si je préfère pas la syntaxe
before==tag_matrice = render_matrix_tag("matrice", 3, 3)==form=={{ tag_matrice }}==
L'idée est que (c'était le cas dans un des exos du wimsathon) on peut avoir à générer toute une liste de tags, notamment pour les drag-drop, de façon diverses, et une fonction render_type_de_tag(paramètres) permet de générer ces tags dans le before avec un max de souplesse. 
Etienne
    Le ‎mercredi‎ ‎17‎ ‎avril‎ ‎2019‎ ‎13‎:‎51‎:‎53‎ ‎CEST, ddoyen <notifications@github.com> a écrit :  
 
 
J'avais essayé de garder la syntaxe générique où toutes les réponses sont dans le dictionnaire answer.

Et comme je voulais disposer d'une réponse détaillée (un booléen qui dit si la liste est bien triée ou pas mais aussi la liste rentrée par l'élève), j'en étais arrivé naturellement à un dictionnaire dans le dictionnaire answer...

Mais effectivement, cette syntaxe est à discuter.

**Comment by ddoyen** [comment]
Effectivement, c'est une syntaxe qui a ses avantages. Mais, dans ce cas, comment va-t-on gérer l'eventuel post-traitement de la réponse de l'élève ?

**Comment by ddoyen** [comment]
Et voilà un essai de syntaxe de `dragdrop`:

https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/dragdrop.pl

**Comment by esandier** [comment]
Oui c'est assez concis. Par contre je ne suis pas pour que la tag s'occupe de corriger et renvoyer un score, parce qu'il peut y avoir plein de situations différentes. Je trouve qu'elle devrait juste dire quel drag a été déposé sur quel drop. Peut-être d'ailleurs qu'il faudrait qu'elle renvoie deux dictionnaires, un dont les clés seraient les drags, l'autre dont les clés seraient les drops, pour parer à toutes les situations. 
Plus généralement, dans mon esprit l'utilisateur lambda utilisera des exos préparés, il y aura donc une surcouche sur cette interface. A ce moment-là, les tags restent un truc relativement bas niveau, qui définit juste un élément d'interface avec l'utilisateur, avec les questions de score, de feedback et tout ce qui concerne l'aspect "exercice" qui seraient traitées dans la surcouche. 
Pour répondre à ta question précédente sur ma suggestion de syntaxe, on peut rajouter une fonction python  qui processe les données du template, comme tu l'as fait dans ton exemple (j'ai pas trop compris comment, mais ça semble assez chouette). 
    Le ‎samedi‎ ‎20‎ ‎avril‎ ‎2019‎ ‎17‎:‎36‎:‎55‎ ‎CEST, ddoyen <notifications@github.com> a écrit :  
 
 
Et voilà un essai de syntaxe de dragdrop:

https://pl.u-pem.fr/filebrowser/option?name=test_pl&path=Yggdrasil/conceptexo/template/dragdrop.pl
